Add option to invert signal polarity

bparks13 opened the pull request:
I have been learning more about DearImGui and how it can be applied to this specific waveform visualizer, particularly in how I can help bring more features from the Open Ephys GUI LFP Viewer into this.

This PR adds a new button at the top of the visualizer that allows the user to invert the polarity of the signal by applying `-1` to the signal.

I did not add the option to invert the polarity as a property of the node, but that can be easily added for consistency, since most of the other parameters in the visualizer are exposed as properties.

bparks13 committed:
Add option to invert signal polarity

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/bonsai-rx/ephys?pullRequest=20) <br/>All committers have signed the CLA.

glopesdev commented on src/Bonsai.Ephys.Design/Decimator.cs:
```diff
@@ -41,11 +41,11 @@ public Decimator(Mat input, int length, int factor, ReduceOperation reduceOperat
         public Depth InputDepth => inputDepth;
 
-        public void Process(Mat input)
+        public void Process(Mat input, bool invertPolarity = false)
```
```suggestion
        public void Process(Mat input, bool invert = false)
```

glopesdev commented on src/Bonsai.Ephys.Design/Decimator.cs:
```diff
@@ -41,11 +41,11 @@ public Decimator(Mat input, int length, int factor, ReduceOperation reduceOperat
             if (conversionBuffer is not null)
             {
-                CV.Convert(input, conversionBuffer);
+                CV.ConvertScale(input, conversionBuffer, invertPolarity ? -1 : 1);
```
```suggestion
                CV.ConvertScale(input, conversionBuffer, invert ? -1 : 1);
```

glopesdev commented on src/Bonsai.Ephys.Design/WaveformVisualizer.cs:
```diff
@@ -51,6 +51,7 @@ public class WaveformVisualizer : BufferedVisualizer
         int colorGrouping = 1;
 
         bool useFixedRange;
+        bool invertSignal = false;
```
```suggestion
        bool invert = false;
```

glopesdev commented on src/Bonsai.Ephys.Design/WaveformVisualizer.cs:
```diff
@@ -130,8 +131,8 @@ public override void Show(object value)
-                decimatorMin.Process(data);
-                decimatorMax.Process(data);
+                decimatorMin.Process(data, invertSignal);
+                decimatorMax.Process(data, invertSignal);
```
```suggestion
                decimatorMin.Process(data, invert);
                decimatorMax.Process(data, invert);
```

glopesdev commented on src/Bonsai.Ephys.Design/WaveformVisualizer.cs:
```diff
@@ -322,6 +323,16 @@ unsafe void MenuWidgets()
+                    invertSignal = !invertSignal;
+                }
+
+                ImGui.PopStyleColor();
```
```suggestion
                ImGui.TableNextColumn();
                var color = invert ? ImGui.GetColorU32(ImGuiCol.ButtonActive) : ImGui.GetColorU32(ImGuiCol.Button);
                ImGui.PushStyleColor(ImGuiCol.Button, color);
                if (ImGui.Button("Invert"u8, buttonSize))
                {
                    invert = !invert;
                }

                ImGui.PopStyleColor();
```

I was also wondering it might be preferable UI-wise to have this button sit next to the `Pause` button since it feels functionally more similar than the color themes (this would be moving the code upward a bit).

glopesdev reviewed (commented):
@bparks13 looks great, thank you for the suggestion! Comments below.

Looking at oscilloscope frontends it looks like the most common name for this feature is simply `invert`. Adopting this would also take up less screen real estate in the button UI, e.g. in case we ever want to place another button above or below it.

Finally, I was wondering whether we might also want to add an `Invert` property to the visualizer settings and operator, to make sure the setting is persistent. In cases where we know the inversion is always required, it would spare people from having to click the button every time the visualizer opens.

bparks13 committed:
Address review comments

- Change name to "invert" to match oscilloscope frontends
- Add property to the visualizer builder so it can be set from the node

bparks13 commented:
@glopesdev I went through and implemented your suggested changes, I agree that keeping it consistent with oscope frontends makes sense.

I did add the property to the visualizer builder so that it can be set from the main Bonsai editor, but I realized that there is no persistence in the other direction. If I open the visualizer and then choose to toggle the Invert option, and then close the visualizer, the original value is still selected in the builder. This is true for all of the current parameters, so it seems to be a feature not a bug, but the other properties use nullable values to prevent overriding values chosen while the visualizer is open. I wasn't sure if using a nullable bool was valid, so I left it this way for now; feel free to modify it as you see fit before merging.

glopesdev committed:
Make invert an optional builder property

This is for consistency with other design properties which set initial
configuration values.

glopesdev committed:
Make invert a checkbox

This helps keep the emphasis on the unique behavior of the pause button
which is tied to a keyboard shortcut.

The "Signal" header is added as a placeholder for a future expandable
section for other signal visualization operations.

glopesdev reviewed (approved):
@bparks13 Looks great, thanks! I have changed the invert control into a checkbox to keep the emphasis on the pause button and added a placeholder `Signal` header for possibly expanding to other operations in the future.
